Shopping Cart

Scharrre97 opened the pull request:
**PT Story:** https://www.pivotaltracker.com/n/projects/1988849

**Changes proposed in this PR:**

* Adding a shopping cart when the user opens the application
* Showing the user that he has added items to his cart


**What I have learned working on this feature:**
We have learn the correct way of adding a shopping cart to an application

**Screenshots:**

![screen shot 2017-03-16 at 10 18 57](https://cloud.githubusercontent.com/assets/25010328/23991387/6e318702-0a39-11e7-9a6a-ea9ea7d95509.png)

Commit:
added shopping_cart test

fabianlundgren committed:
Added shopping_cart controller and add to cart button

fabianlundgren committed:
Added action to controller

AmberWilkie committed:
Add acts_as_shopping_cart to gemfile

AmberWilkie committed:
Install acts_as_shopping_cart models and db tables

AmberWilkie committed:
Set up cucumber step definition for seeing an image in the cart

casperjureskog commented on features/step_definitions/basic_steps.rb:
```diff
@@ -10,4 +10,8 @@
+end
+
+When(/^I click "([^"]*)"$/) do |button|
+  click_link_or_button button
```
Possible merge conflict later

Scharrre97 replied:
Good point. We need to know that.

Commit:
added first rspec for Shopping cart

fabianlundgren committed:
Fixed the add to cart test

Commit:
added image to shopping cart

Scharrre97 committed:
Merge branch 'develop' into shopping_cart

sigu commented on spec/factories/shopping_cart_items.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :shopping_cart_item do
+    
+  end
```
The factories are empty

sigu commented on spec/factories/shopping_carts.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :shopping_cart do
```
Same empty factory

Scharrre97 replied:
In the schema there's only created_at and updated_at.

tochman replied:
I don't have a problem with this. We will expand this factory as we add more attributes to ShoppingCart. Like associating it with a User, etc.

sigu commented on spec/models/shopping_cart_item_spec.rb:
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe ShoppingCartItem, type: :model do
+  describe 'validations' do
```
Where are the tests for the necessary validations? If not required then lets not have empty method

sigu commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,15 @@
+class ShoppingCartController < ApplicationController
+
```
unnecessary empty line

Commit:
added factory test and fix PR comments

Commit:
Merge branch 'shopping_cart' of https://github.com/Scharrre97/photo_market into shopping_cart

fabianlundgren committed:
Fixed a typo and added the file name into background

fabianlundgren committed:
Merge branch 'shopping_cart' of https://github.com/Scharrre97/photo_market into shopping_cart

tochman commented on features/shopping_cart.feature:
```diff
@@ -0,0 +1,13 @@
+Scenario: I add a image to my shopping cart
+  Given I am on the index page
+  When I click "Add to cart"
+  Then I should see "1" image in my shopping cart
```
And then what?

tochman replied:
Also, mind the code formatting

tochman commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,13 @@
+class ShoppingCartController < ApplicationController
+    def add_to_cart
+
```
Remove empty lines plz

tochman commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,13 @@
+class ShoppingCartController < ApplicationController
+    def add_to_cart
+
+      if cookies['cart_id']
```
Where is this cookie getting set?

tochman commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,13 @@
+    def add_to_cart
+
+      if cookies['cart_id']
+        @cart = ShoppingCart.find_or_create_by(id: session[:cart_id])
```
Why are you using an instance variable and not a local one?

tochman replied:
Also, on line 4 you are looking for a cookie and here you are getting it from a `session`, why?

tochman commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,13 @@
+        @cart = ShoppingCart.create
+      end
+
+      @image = Image.find(params[:image_id])
```
Again, why an instance variable?

tochman commented on app/controllers/shopping_cart_controller.rb:
```diff
@@ -0,0 +1,13 @@
+
+      @image = Image.find(params[:image_id])
+      @cart.add(@image, @image.price)
+    end
```
What happens once all this is being set? What view will be rendered for the user?

tochman commented on app/views/images/index.html.erb:
```diff
@@ -3,10 +3,19 @@
+        <%= hidden_field_tag(:image_id, image.id) %>
+        <%= hidden_field_tag(:image_price, image.price) %>
+        <%= submit_tag("Add to cart") %>
+      <% end %>
```
This form should be placed inside the div below. That div is meant to be a wrapper for everything related to the specific image we are displaying atm. That's why we gave it an class of  `"image-<%=image.id%>"`

tochman commented on config/routes.rb:
```diff
@@ -1,4 +1,9 @@
   devise_for :users
+
   root controller: :images, action: :index
+  post '/add_to_cart', controller: :shopping_cart, action: :add_to_cart, as: :add_to_cart
```
Why `as: :add_to_cart`?

tochman reviewed (dismissed):
I have a few questions. I'm a bit worried that you allowed yourself to be inspired by some ready made solution rather than driving your development with tests. 
I can work with you on this one but the way this look atm feels a bit weird. Please be mindful of the Acceptance-Unit Test Cycle flow.

tochman committed:
refactort routes for shopping_cart to more restful standard
renames controller to follow rails standard
refactores the view
adds helper method to display information about item count
adds flash on successful addition of image to cart
refactors and adds new scenario to feature

tochman committed:
merges with upstream and fixex conflicts
moves item count to header

tochman committed:
moves #create_cart to application_controller

diraulo commented on app/controllers/application_controller.rb:
```diff
@@ -1,10 +1,15 @@
   end
+
+  def create_cart
+    session[:cart_id] = ShoppingCart.create.id unless session[:cart_id]
```
This could be written as
```ruby
session[:card_id] ||= ShoppingCart.create.id
```

tochman replied:
@Scharrre97 @fabianlundgren Please refactor this as @diraulo suggests.

tochman replied:
On a second thought, are you sure this does the same thing @diraulo ? I have not tested it but I feel this will not cut it

diraulo replied:
```ruby
session[:card_id] ||= ShoppingCart.create.id
# is equivalent to 
session[:card_id] = session[:cart_id] || ShoppingCart.create.id
```

This means if `session[:cart_id]` evaluates to `nil` or `false` we create a new `ShoppingCart` and get its `id` assigned to `session[:cart_id]`

Every other call to this method will not override the value of `session[:cart_id]` unless it was cleared somewhere else in the app (ie: shopping cart destroyed)

tochman replied:
We can debate this or you can try by adding this code and running the tests.

Scharrre97 replied:
We will keep it the way it is, because it works and we need to move on.

diraulo replied:
I understand the need to move on @Scharrre97, but the only reason I made this comment was to expose you guys to a different approach to do the exact same thing. 

All I expected was for you guys to at least take the time to try it out and maybe learn 1 new thing out of this. Don't forget that as much as the objective is to deliver the feature, the main thing is for you to learn something out of this experience.

This code works as it is and I have no objection to merging it.

tochman replied:
I would also suggest that you at least try the approach suggested by @diraulo I did that and I would like you to see the results.

diraulo commented on app/controllers/application_controller.rb:
```diff
@@ -1,10 +1,15 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   before_action :configure_permitted_parameters, if: :devise_controller?
+  before_action :create_cart
```
Something about this doesn't feel right to me... Why do we want to call `create_cart` on literally every action we do in this app? What was the reason for moving it out of the ShoppingCartController?

tochman replied:
In this set up we are creating a cart as soon as a visitor opens the app. That allows us to have the cart info persisted even if the user chooses to leave the application. Next time he visits the site the cart will still be there.

tochman replied:
And if he logs in we can update the cart with his info (not part of this story though - there's no association ShoppingCart belongs_to User yet)

diraulo replied:
But that can also be done the first time he adds an item to the shopping cart. If we don't have any cart created in session then we create one at that time. Every controller inherits from the `ApplicationController` and that means this method will be executed for every action we do in the application. It probably has no serious implication in this particular application but I'd be cautious about this kind of things.

AmberWilkie replied:
Yes why wouldn't we do this when he adds the first item to the cart? That's how I set up this feature.

tochman replied:
This was my proposal but I'm open to other implementations. Go ahead and work with @Scharrre97 and @fabianlundgren 
As to the performance issues this method is only hitting the db once per session. All other times it only checks the session and that is not performance heavy.

tochman replied:
Will there be any action on this? @Scharrre97 @fabianlundgren @AmberWilkie or @diraulo ?

Scharrre97 replied:
We don't see any problem with this, both ways make sense to us so we will keep it the way it is. Let's move on to the next feature.

diraulo replied:
@tochman This specific case is not performance heavy as I pointed out in my previous comment so I'm okay with letting this go. I just thought it was necessary to point out the implications of doing something like this.

diraulo commented on app/controllers/images_controller.rb:
```diff
@@ -1,5 +1,9 @@
 class ImagesController < ApplicationController
+
```
What's up with all the empty lines?

tochman replied:
Just a fuck up on my side probably....

tochman committed:
merging with upstream

casperjureskog commented on app/controllers/images_controller.rb:
```diff
@@ -1,5 +1,9 @@
   def index
     @images = Image.all
   end
+
```
Please remove empty lines

Commit:
fix .gitignore and remove empty lines

tochman commented:
How's that PR update going @Scharrre97 ?

casperjureskog commented:
Its looks good cant see any problems

fottander commented:
Looks good to me.

diraulo commented:
@tochman you still need to approve this though since you requested some changes

diraulo reviewed (approved):
:shipit:

tochman commented:
I was in a meeting. Thank you for moving mt review out of the way @diraulo
